### ISSUE TITLE
Keras backend functions not working as intended?

### ISSUE THREAD

**Issue by korr4k** [tensorflow/tensorflow#35330]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution: 4.4.0-18362-Microsoft
- TensorFlow installed from: Anaconda default source
- TensorFlow version: 1.15
- Python version: 3.7.5
- CUDA/cuDNN version: 10.0
- GPU model and memory: GeForce RTX 2060

**Describe the current behavior**

I'm trying to implement a custom loss function based on a custom accuracy function that I'm already using to evaluate my model predictions on the test dataset. The conversion can't be 1:1 because I use numpy "greater" and "equal" functions that are not differentiable. I created thefore custom functions that approximate the latters but their behavior has some problems

**Describe the expected behavior**

In particular I can test if everything is fine by comparing the results obtained by my original custom accuracy f. and the new loss f. given the same input (my input are tensorflow predictions, I just inglobe them inside K.constant to convert them in tensors). What I noticed is that this line of code

```
eps = sys.float_info.epsilon
return 0.5*(y + 5 + K.sqrt(K.pow(y-5,2) + eps))
```

is problematic. In particular y is an array of float32 values in [1:10] range and the returned array, let's call it 'ret', should have ret[i]=max(5,y[i]) but sometimes the value of **5** becomes **4.9999995** instead. The next portion of my code is based on how many **5** are present and therefore I can't ignore this problem.

The fact is that, let's say a problematic index is 'w' so that ret[w]=4.9999995 instead of 5, if I use the same code with y now equal to only y[w] the returned array is correctly 5. This means that somehow if y is a batch of predictions and not just one something isn't working. This should not be the case because both K.sqrt and K.pow works element wise, it should not matter if y is an array of 1 or multiple values 

Out of almost 20k predictions, around 1k have this same problem and it is deterministic (always the same are problematic). I also tried to use:

```
eps = sys.float_info.epsilon
return 0.5*(y + 5 + np.sqrt(np.pow(y-5,2) + eps))
```

and the problem is gone thefore it is related to Keras backend.

Last info, I tried to use also:

```
eps = sys.float_info.epsilon
return tf.math.ceil(0.5*(y + 5 + K.sqrt(K.pow(y-5,2) + eps)))
```

but this completely ruins the returned value, sometimes real numbers such as 4.5 are rounded to 6 instead of 5

If more informations are needed I can provide them

**Comment by amahendrakar** [comment]
@korr4k,
In order to expedite the trouble-shooting process, please provide the complete code snippet to reproduce the issue reported here. Thanks!

**Comment by korr4k** [comment]
```Python
from tensorflow.compat.v1.keras import backend as K
import numpy as np
import sys
eps = sys.float_info.epsilon

def compute_wms_keras(y):
    # Numpy vector of weights
    w = np.arange(1,11).astype('float32')
    w = np.reshape(w, (10,1))
    w = tf.convert_to_tensor(w, dtype=tf.float32)

    # wms of y by w
    return K.dot(y, w)

def greater_approx(y, threshold):    
    return 0.5*(y + threshold + K.sqrt(K.pow(y-threshold,2) + eps))

def gaussian(x, mu):
    sig=eps
    return K.exp(-K.pow((x - mu),2) / (2.0 * sig*sig))

def custom_accuracy_keras(y, y_first):

    # Compute the wms of both y and y_first
    y_wms = compute_wms_keras(y)
    y_first_wms = compute_wms_keras(y_first)

    # greater_approx and gaussian are the approximation functions of "greater"
    # and "equal" respectively
    greater_approx_y_wms = greater_approx(y_wms, 5.0)
    greater_approx_y_first_wms = greater_approx(y_first_wms, 5.0)
    
    gaussian_y_wsm = gaussian(greater_approx_y_wms, 5.0)
    gaussian_y_first_wms = gaussian(greater_approx_y_first_wms, 5.0)
    
    matches = gaussian(gaussian_y_wsm, gaussian_y_first_wms)
    
    # matches at this point is an array of 0 and 1, 0 when the accuracy test
    # failed and 1 otherwise
    return y_wms, y_first_wms, greater_approx_y_wms, greater_approx_y_first_wms, gaussian_y_wsm, gaussian_y_first_wms, matches, K.mean(matches)
```

Both input y and y_first are array of shape (x,10), where x is the number of elements in the dataset, with "compute_wms" they are converted in the shape of (x,1). Note that "custom_accuracy_keras" should return only K.mean(matches) but to identify where the problem it is temporary changed

The orginal numpy accuracy function, the one correct, is:

```Python
import numpy as np

def compute_wms(y):
    # Numpy vector of weights
    w = np.arange(1,11).astype('float32')
    w = np.reshape(w, (10,1))

    # wms of y by w
    return np.dot(y, w)

def custom_accuracy(y, y_first):

    # Compute the wms of both y and y_first
    y_wms = compute_wms(y)
    y_first_wms = compute_wms(y_first)

    # Are the obtained values greater than 5?
    y_true_bool = np.greater(y_wms, 5)
    y_pred_bool = np.greater(y_first_wms, 5)

    # The results of the previous tests are compared
    matches = np.equal(y_true_bool, y_pred_bool).astype('float32')

    # The mean of the previous test is returned
    return y_wms, y_first_wms, matches, np.mean(matches)
```

Note same as above for the returned value.

To test the code, after loading y and y_first numpy array, I use this code:

```Python
from tensorflow.compat.v1.keras import backend as K
import tensorflow as tf
sess = tf.Session()
import numpy as np

correct_results = custom_accuracy(y,y_first)

# Originally I used K.constant instead of tf.convert_to_tensor but I think it is the same
y_wms, y_first_wms, greater_approx_y_wms, greater_approx_y_first_wms, gaussian_y_wsm, gaussian_y_first_wms, matches, meatches_mean = custom_accuracy_keras(tf.convert_to_tensor(y), tf.convert_to_tensor(y_first))

y_wms = y_wms.eval(session=sess)
y_first_wms = y_first_wms.eval(session=sess)
greater_approx_y_wms = greater_approx_y_wms.eval(session=sess)
greater_approx_y_first_wms = greater_approx_y_first_wms.eval(session=sess)
gaussian_y_wsm = gaussian_y_wsm.eval(session=sess)
gaussian_y_first_wms = gaussian_y_first_wms.eval(session=sess)
matches = matches.eval(session=sess)
meatches_mean = meatches_mean.eval(session=sess)                                                                                                                                            

errors = []
indexes= []
        
for i in range(matches.size):            
    if(matches[i][0]!=correct_results[2][i]):
        errors.append((y_wms[i][0],
                        y_first_wms[i][0],
                        greater_approx_y_wms[i][0],
                        greater_approx_y_first_wms[i][0],
                        gaussian_y_wsm[i][0],
                        gaussian_y_first_wms[i][0],
                        matches[i][0]))
        indexes.append(i)
```
At this point in "errors" I have all the data relative to input with problems and in "indexes" their relative position

To test what I reported the code is:

```Python
test = []
        
for i in indexes[:10]:
    y_wms, y_first_wms, greater_approx_y_wms, greater_approx_y_first_wms, gaussian_y_wsm, gaussian_y_first_wms, matches, meatches_mean = custom_accuracy_keras(tf.convert_to_tensor(y[i].reshape((1, 10)), dtype=tf.float32), tf.convert_to_tensor(y_first[i].reshape((1, 10)), dtype=tf.float32))
            
    test.append((y_wms[0].eval(session=sess)[0],
                 y_first_wms[0].eval(session=sess)[0],
                 greater_approx_y_wms[0].eval(session=sess)[0],
                 greater_approx_y_first_wms[0].eval(session=sess)[0],
                 gaussian_y_wsm[0].eval(session=sess)[0],
                 gaussian_y_first_wms[0].eval(session=sess)[0],
                 matches[0].eval(session=sess)[0]))
```

And now the first 10 (I use only the first 10 for simplicity) n tuples of "errors" should be the same as "test" but they aren't and most importantly "test" has the correct results.
Note: the colomuns where the **5** becomes **4,9999995** but also the first two don't have the exact same values! Maybe the two things are related

[Input.zip](https://github.com/tensorflow/tensorflow/files/3994594/Input.zip): this is a zip with both the numpy array y and y_first

**Comment by amahendrakar** [comment]
Was able to reproduce the issue. Please find the [Gist](https://colab.sandbox.google.com/gist/amahendrakar/090aeeccec85f1a475bfd90f3f241d83/35330.ipynb) here. Thanks!

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
